pkg/appclient/appclient.go

```diff
@@ -0,0 +1,159 @@
+	keyName string
+	keyring keyring.Keyring
+
+	// clientCtx is the client context for the application.
+	// It is used to query for the application's account to unmarshal the supplier's account
+	// and get the public key to verify the relay response signature.
+	clientCtx sdkclient.Context
+
+	// appAddress is the address of the application that this app client is for.
+	appAddress string
+
+	// sessionQuerier is the querier for the session module.
+	// It used to get the current session for the application given a requested service.
+	sessionQuerier sessiontypes.QueryClient
+
+	// accountQuerier is the querier for the account module.
+	// It is used to get the the supplier's public key to verify the relay response signature.
+	accountQuerier accounttypes.QueryClient
+
+	// blockClient is the client for the block module.
+	// It is used to get the current block height to query for the current session.
+	blockClient blocktypes.BlockClient
+
+	// server is the HTTP server that will be used capture application requests
+	// so that they can be signed and relayed to the supplier.
+	server *http.Server
+}
+
+func NewAppClient(
+	clientCtx sdkclient.Context,
+	keyName string,
+	keyring keyring.Keyring,
+	applicationEndpoint *url.URL,
+	blockClient blocktypes.BlockClient,
+) *appClient {
+	sessionQuerier := sessiontypes.NewQueryClient(clientCtx)
+	accountQuerier := accounttypes.NewQueryClient(clientCtx)
+
+	return &appClient{
+		clientCtx:      clientCtx,
+		keyName:        keyName,
+		keyring:        keyring,
+		sessionQuerier: sessionQuerier,
+		accountQuerier: accountQuerier,
+		blockClient:    blockClient,
+		server:         &http.Server{Addr: applicationEndpoint.Host},
+	}
+}
+
+// Start starts the application server and blocks until the context is done
+// or the server returns an error.
+func (app *appClient) Start(ctx context.Context) error {
+	// Get and populate the application address from the keyring.
+	keyRecord, err := app.keyring.Key(app.keyName)
+	if err != nil {
+		return err
+	}
+
+	accAddress, err := keyRecord.GetAddress()
+	if err != nil {
+		return err
+	}
+
+	app.appAddress = accAddress.String()
+
+	// Shutdown the HTTP server when the context is done.
+	go func() {
+		<-ctx.Done()
+		app.server.Shutdown(ctx)
+	}()
+
+	// Start the HTTP server.
+	return app.server.ListenAndServe()
+}
+
+// Stop stops the application server and returns any error that occurred.
+func (app *appClient) Stop(ctx context.Context) error {
+	return app.server.Shutdown(ctx)
+}
+
+// ServeHTTP is the HTTP handler for the application server.
+// It captures the application request, signs it, and sends it to the supplier.
+// After receiving the response from the supplier, it verifies the response signature
+// before returning the response to the application.
+// The serviceId is extracted from the request path.
+func (app *appClient) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
```

Comment:
Confirming that this needs to be publically exposed?

Reply:
We are complying to the `http.Server` interface.